src/random.rs

```diff
@@ -0,0 +1,722 @@
+                lb.as_ptr(),
+                ub.as_ptr(),
+                shape.as_ptr(),
+                shape.len(),
+                T::DTYPE.into(),
+                key.as_ptr(),
+                stream.as_ptr(),
+            )
+        };
+        Ok(Array::from_ptr(c_array))
+    }
+}
+
+/// Sample from the standard Gumbel distribution.
+///
+/// The values are sampled from a standard Gumbel distribution
+/// which CDF `exp(-exp(-x))`.
+///
+/// ```rust
+/// let key = mlx_rs::random::key(0);
+///
+/// // generate a single Float with Gumbel distribution
+/// let value = mlx_rs::random::gumbel::<f32>(None, &key).unwrap().item::<f32>();
+///
+/// // generate an array of Float with Gumbel distribution in shape [10, 5]
+/// let array = mlx_rs::random::gumbel::<f32>(&[10, 5][..], &key);
+/// ```
+#[default_device]
+pub fn gumbel_device<'a, T: ArrayElement>(
+    shape: impl Into<Option<&'a [i32]>>,
+    key: impl Into<Option<&'a Array>>,
+    stream: StreamOrDevice,
+) -> Result<Array, Exception> {
+    let shape = shape.into().unwrap_or(&[]);
+    let key = key.into().map_or_else(
+        || {
+            let mut state = state().lock().unwrap();
+            state.next()
+        },
+        |key| key.clone(),
+    );
+
+    unsafe {
+        let c_array = try_catch_c_ptr_expr! {
+            mlx_sys::mlx_random_gumbel(
+                shape.as_ptr(),
+                shape.len(),
+                T::DTYPE.into(),
+                key.as_ptr(),
+                stream.as_ptr(),
+            )
+        };
+        Ok(Array::from_ptr(c_array))
+    }
+}
+
+/// Sample from a categorical distribution.
+///
+/// The values are sampled from the categorical distribution specified by
+/// the unnormalized values in `logits`.   If the `shape` is not specified
+/// the result shape will be the same shape as `logits` with the `axis`
+/// dimension removed.
+///
+/// ```rust
+/// let key = mlx_rs::random::key(0);
+///
+/// let logits = mlx_rs::Array::zeros::<u32>(&[5, 20]).unwrap();
+///
+/// // produces Array of u32 shape &[5]
+/// let result = mlx_rs::random::categorical(&logits, None, None, None, &key);
+/// ```
+///
+/// - Parameters:
+///     - logits: The *unnormalized* categorical distribution(s).
+#[default_device]
+pub fn categorical_device<'a>(
+    logits: &Array,
+    axis: impl Into<Option<i32>>,
+    shape: impl Into<Option<&'a [i32]>>,
+    num_samples: impl Into<Option<i32>>,
+    key: impl Into<Option<&'a Array>>,
+    stream: StreamOrDevice,
+) -> Result<Array, Exception> {
+    let axis = axis.into().unwrap_or(-1);
+
+    let key = key.into().map_or_else(
+        || {
+            let mut state = state().lock().unwrap();
+            state.next()
+        },
+        |key| key.clone(),
+    );
+
+    let shape = shape.into();
+    let num_samples = num_samples.into();
+
+    // if we have both shape and num_samples, we need to error out
+    if shape.is_some() && num_samples.is_some() {
```

Comment:
I think this can be solved by introducing a new type?

```rust
pub enum ShapeOrCount {
    Shape(&[i32]),
    Count(i32)
}
```

Then the arg takes `shape_or_count: impl Into<Option<ShapeOrCount>>` (or `IntoOption<ShapeOrCount>`).